### PR TITLE
Add support for price functions on shipping zone rate tiers

### PR DESCRIPTION
**Pull request by petehayes102**
Note that when I ran `make docs`, it introduced a tonne of changes that are not related to my change. I'm happy to add the changes to this PR but wanted to confirm with the maintainers first.

### CODE REVIEW

File: commercetools/resource_shipping_zone_rate.go [hunk 1]
```diff
@@ -264,28 +283,43 @@ func unmarshallShippingRatePriceTiers(d *schema.ResourceData) ([]platform.Shippi
 	for _, priceTier := range values.([]interface{}) {
 		tierMap := priceTier.(map[string]interface{})
 
-		priceMap := tierMap["price"].([]interface{})[0].(map[string]interface{})
-		price := platform.Money{
-			CurrencyCode: priceMap["currency_code"].(string),
-			CentAmount:   priceMap["cent_amount"].(int),
+		var price *platform.Money
+		priceList := tierMap["price"].([]interface{})
+		if len(priceList) > 0 {
```
Comment (mvantellingen) [inline commercetools/resource_shipping_zone_rate.go:288]: There is a utility function available to achieve this:

`priceMap, err := elementFromSlice(priceTier, "price")`

File: commercetools/resource_shipping_zone_rate.go [hunk 2]
```diff
@@ -264,28 +283,43 @@ func unmarshallShippingRatePriceTiers(d *schema.ResourceData) ([]platform.Shippi
...
 		case string(platform.ShippingRateTierTypeCartClassification):
 			tiers = append(tiers, platform.CartClassificationTier{
 				Value: tierMap["value"].(string),
-				Price: price,
+				Price: *price,
 			})
 		case string(platform.ShippingRateTierTypeCartScore):
+			var function *platform.PriceFunction
+			functionList := tierMap["price_function"].([]interface{})
+			if len(functionList) > 0 {
```
Comment (mvantellingen) [inline commercetools/resource_shipping_zone_rate.go:311]: Same as above, you can leverage `elementFromSlice()`

**Review by mvantellingen — CHANGES_REQUESTED** [review]

**Comment by mvantellingen** [comment]
Looks good! Two small changes to simplify the code. Note that i'm more and more doubting if the added nesting of resources makes it better. 

So instead of 

```hcl
        shipping_rate_price_tier {
            type               = "CartScore"
            score              = 30
            price_function {
                function      = "x + 1"
                currency_code = "%[3]s"
            }
```

flatten it as 
```hcl
        shipping_rate_price_tier {
            type               = "CartScore"
            score              = 30
            price_function_expression = "x + 1"
            price_function_currency_code = "%[3]s"
}
```

But that might be food for thought for the future

**Comment by jpuffen** [comment]
any update on this?

**Comment by mvantellingen** [comment]
I was in the process of rebasing this but stumbled upon a question.

We now have

```hcl
  shipping_rate_price_tier {
    type  = "CartScore"
    score = 10

    price {
      cent_amount   = 5000
      currency_code = "EUR"
    }
  }
```

and for the price function you added

```hcl
  shipping_rate_price_tier {
    type  = "CartScore"
    score = 10

    price_function {
      function   = "x + 1"
      currency_code = "EUR"
    }
  }
```

Is there a reason why we don't use:


```hcl
  shipping_rate_price_tier {
    type  = "CartScore"
    score = 10

    price {
      function   = "x + 1"
      currency_code = "EUR"
    }
  }
```

**Comment by petehayes102** [comment]
Sorry for the delay here.

@jpuffen I've been pulled onto other projects and have never written Go before/since, so amending the code will take a bit of trial and error. I'll try and revisit this during the week.

@mvantellingen Yeah good question. I think I was trying to follow what I thought was the convention, but your suggestion is cleaner. Happy to bundle that in with the other changes.

**Comment by mvantellingen** [comment]
Cool, i've pushed the rebase I did, you can find it here: https://github.com/labd/terraform-provider-commercetools/tree/petehayes102-shipping-rate-pricefn

**Comment by petehayes102** [comment]
@mvantellingen I just pushed the rebased changes, but it seems to have introduced a bunch of unintended changes...

**Comment by petehayes102** [comment]
Ok that's better. I rebased again from `main` and now it's back to just my changeset.

**Comment by mvantellingen** [comment]
I've merged this but ran in a couple of issues. I think my suggestion of using a single `price` block with either `function` or `cent_amount` is not the correct one :-) since it conflicts with how commercetools behaves.

I'll fix it this week

**Comment by mvantellingen** [comment]
Fixed, will be in v1.5.0

Thanks!

**Comment by petehayes102** [comment]
No dramas mate. Glad we got there in the end! :)

**Pull request merged**